python/ray/data/_internal/compute.py

```diff
@@ -80,32 +81,66 @@ def _apply(
             return block_list
 
         blocks = block_list.get_blocks_with_metadata()
+        # Bin blocks by target block size.
+        if target_block_size is not None:
+            binned_blocks = []
+            curr_bin = []
+            curr_bin_size = 0
+            for b, m in blocks:
+                num_rows = m.num_rows
+                if num_rows is None:
+                    num_rows = float("inf")
```

Comment:
Log a warning that binning isn't possible?

Reply:
Hmm that might be noisy and not very actionable, e.g. for datasources that don't advertise the number of rows at metadata fetch time (such as CSV and JSON files).

Reply:
Ok, we can see if it ends up a point of confusion